src/crystal/system/unix/io_uring.cr

```diff
@@ -0,0 +1,569 @@
+
+      if mem.address.to_i64! < 0
+        err = Errno.new(-mem.address.to_i64!.to_i)
+        raise RuntimeError.from_os_error("Cannot allocate completion queue", err)
+      end
+
+      @completion_queue_mmap = mem
+    end
+
+    mem = Syscall.mmap(Pointer(Void).null, sq_entries_size,
+      Syscall::Prot::READ | Syscall::Prot::WRITE, Syscall::Map::SHARED | Syscall::Map::POPULATE,
+      @fd, Syscall::IORING_OFF_SQES)
+
+    if mem.address.to_i64! < 0
+      err = Errno.new(-mem.address.to_i64!.to_i)
+      raise RuntimeError.from_os_error("Cannot allocate submission entries", err)
+    end
+
+    @submission_entries = mem.as(Syscall::IoUringSqe*)
+
+    @submission_queue = SubmissionQueue.new(@fd, @submission_queue_mmap, @params.sq_off)
+    @completion_queue = CompletionQueue.new(@fd, @completion_queue_mmap, @params.cq_off)
+  end
+
+  def finalize
+    if @submission_queue_mmap && @submission_queue_mmap == @completion_queue_mmap
+      Syscall.munmap(@submission_queue_mmap, {sq_size, cq_size}.max)
+    end
+
+    if @submission_queue_mmap
+      Syscall.munmap(@submission_queue_mmap, sq_size)
+    end
+
+    if @completion_queue_mmap
+      Syscall.munmap(@completion_queue_mmap, cq_size)
+    end
+
+    if @completion_queue_mmap
+      Syscall.munmap(@completion_queue_mmap, sq_entries_size)
+    end
+
+    if @fd > 0
+      Syscall.close(@fd)
+    end
+  end
+
+  private def sq_size
+    Syscall::SizeT.new(@params.sq_off.array + @params.sq_entries * sizeof(UInt32))
+  end
+
+  private def cq_size
+    Syscall::SizeT.new(@params.cq_off.cqes + @params.cq_entries * sizeof(Syscall::IoUringCqe))
+  end
+
+  private def sq_entries_size
+    Syscall::SizeT.new(@params.sq_entries * sizeof(Syscall::IoUringSqe))
+  end
+
+  # Data pointed by the `user_data` field
+  private struct UserData
+    property res = 0
+    property fiber = ::Fiber.current
+  end
+
+  # Obtains one submission entry, populates, and submits it. When the completion
+  # event arrives the current Fiber will be enqueued for execution.
+  # The caller of this method MUST either reschedule the current Fiber or
+  # resume into another Fiber.
+  private def submit!(opcode : Syscall::IoUringOp, user_data_pointer : Pointer(UserData))
+    # Obtains one free index from the submission queue for our use. If there is none
+    # available then we should call into the kernel to process what is there. Retry
+    # until the kernel has consumed at least one request so that we can write ours.
+    until (index = @submission_queue.consume_free_index) != UInt32::MAX
+      process_completion_events(blocking: false)
+    end
+
+    # The submission queue just stores indices into the `submission_entries` array.
+    sqe = @submission_entries + index
+
+    # This submission queue entry is reused! Clean it up.
+    Intrinsics.memset(sqe, 0_u8, sizeof(Syscall::IoUringSqe), false)
+
+    # Populate fields and submit the index into the submission queue.
+    yield sqe
```

Comment:
I added support for a linked timeout. It is handled as a special case since the only kind of linked request we need for implementing the scheduler/eventloop is a read/write timeout. This interface won't support free-form linked requests because there is no need.